GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs

```diff
@@ -163,35 +227,39 @@ public virtual void AutoAttackHit(IAttackableUnit target)
             }
 
             var damage = Stats.AttackDamage.Total;
-            if (_isNextAutoCrit)
+            if (IsNextAutoCrit)
             {
                 damage *= Stats.CriticalDamage.Total;
             }
 
-            var onAutoAttack = _game.ScriptEngine.GetStaticMethod<Action<IAttackableUnit, IAttackableUnit>>(Model, "Passive", "OnAutoAttack");
-            onAutoAttack?.Invoke(this, target);
-
             target.TakeDamage(this, damage, DamageType.DAMAGE_TYPE_PHYSICAL,
                 DamageSource.DAMAGE_SOURCE_ATTACK,
-                _isNextAutoCrit);
+                IsNextAutoCrit);
         }
 
-        /// <summary>
-        /// Sets this AI's current auto attack to the given auto attack. *NOTE*: Will be deprecated when spells are fully implemented.
-        /// </summary>
-        /// <param name="newAutoAttackSpellData">Auto attack spell data to use.</param>
-        public void ChangeAutoAttackSpellData(ISpellData newAutoAttackSpellData)
+        public override bool CanMove()
         {
-            AaSpellData = newAutoAttackSpellData;
+            // False if any are true.
+            return !(IsDead || MoveOrder == OrderType.CastSpell
+                    // TODO: Remove these and implement them as buffs, then just check the BuffType here.
+                    || HasCrowdControl(CrowdControlType.AIRBORNE)
+                    || HasCrowdControl(CrowdControlType.ROOT)
+                    || HasCrowdControl(CrowdControlType.STASIS)
+                    || HasCrowdControl(CrowdControlType.STUN)
+                    || HasCrowdControl(CrowdControlType.SNARE));
         }
 
         /// <summary>
-        /// Sets this AI's current auto attack to the given auto attack. *NOTE*: Will be deprecated when spells are fully implemented.
+        /// Whether or not this AI is able to cast spells.
         /// </summary>
-        /// <param name="newAutoAttackSpellDataName">Name of the auto attack to use.</param>
-        public void ChangeAutoAttackSpellData(string newAutoAttackSpellDataName)
+        public bool CanCast()
+        {
+            return !HasCrowdControl(CrowdControlType.STUN) && !HasCrowdControl(CrowdControlType.SILENCE);
+        }
+
+        public bool CanLevelUpSpell(ISpell s)
         {
-            AaSpellData = _game.Config.ContentManager.GetSpellData(newAutoAttackSpellDataName);
+            return CharData.SpellsUpLevels[s.CastInfo.SpellSlot][s.CastInfo.SpellLevel] <= Stats.Level;
```

Comment:
Why is it 2D array?

Reply:
Each slot can have its own level requirement in order to be leveled up. In particular, ultimates require a high level gap between level ups.